cmd/bacalhau/validate.go

```diff
@@ -0,0 +1,129 @@
+package bacalhau
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/filecoin-project/bacalhau/pkg/model"
+	"github.com/filecoin-project/bacalhau/pkg/util/templates"
+	"github.com/spf13/cobra"
+
+	"gopkg.in/yaml.v3"
+	"k8s.io/kubectl/pkg/util/i18n"
+	convert "sigs.k8s.io/yaml"
+
+	"github.com/invopop/jsonschema"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+var (
+	validateLong = templates.LongDesc(i18n.T(`
+		validate a job from a file
+
+		JSON and YAML formats are accepted.
+`))
+
+	//nolint:lll // Documentation
+	validateExample = templates.Examples(i18n.T(`
+		# validate a job using the data in job.yaml
+		bacalhau validate ./job.yaml
+`))
+
+	// Set Defaults (probably a better way to do this)
+	OV = NewValidateOptions()
+
+	// For the -f flag
+)
+
+type ValidateOptions struct {
+	Filename string // Filename for job (can be .json or .yaml)
+}
+
+func NewValidateOptions() *ValidateOptions {
+	return &ValidateOptions{
+		Filename: "",
+	}
+}
+
+var validateCmd = &cobra.Command{
+	Use:     "validate",
+	Short:   "validate a job using a json or yaml file.",
+	Long:    validateLong,
+	Example: validateExample,
+	Args:    cobra.MinimumNArgs(0),
+	RunE: func(cmd *cobra.Command, cmdArgs []string) error { //nolint:unparam // incorrect that cmd is unused.
+
+		if len(cmdArgs) == 0 {
+			_ = cmd.Usage()
+			return fmt.Errorf("no filename specified")
+		}
+		OV.Filename = cmdArgs[0]
+
+		fileextension := filepath.Ext(OV.Filename)
+		fileContent, err := os.Open(OV.Filename)
+
+		if err != nil {
+			return fmt.Errorf("could not open file '%s': %s", OV.Filename, err)
+		}
+
+		byteResult, err := io.ReadAll(fileContent)
+
+		if err != nil {
+			return err
+		}
+
+		jobSpec := &model.JobSpec{}
+
+		if fileextension == ".json" {
```

Comment:
I’ve seen this pattern a couple of times in the code and it’s always bugged me (lots of things don’t have file extensions, like stdin). I wonder if a solution like this - https://stackoverflow.com/a/68159730/4322 - is more elegant. Regardless, we should use generics, and move to a function